File: lab2/src/Algorithm.cpp
#include "Algorithm.h"

Algorithm::Algorithm(const Matrix &a, const Matrix &b, const int &m, const int &n, const int &q) : a(a), b(b), m(m),
    n(n), q(q) {
    if (m < 1 || n < 1 || q < 1) {
        throw std::invalid_argument("Algorithm: invalid arguments");
    }
    if (a.size() != m || b.size() != n) {
        throw std::invalid_argument("Algorithm: invalid arguments");
    }
    for (auto i = 0; i < m; ++i) {
        if (a[i].size() != n) {
            throw std::invalid_argument("Algorithm: invalid arguments");
        }
    }
    for (auto i = 0; i < n; ++i) {
        if (b[i].size() != q) {
            throw std::invalid_argument("Algorithm: invalid arguments");
        }
    }
}

File: lab6/src/main/java/algorithms/Algorithm.java
package algorithms;

import org.jetbrains.annotations.NotNull;

public abstract class Algorithm {
    static final double EPSILON = 1e-6;
    static final int DAYS = 80;

    public double solve(double @NotNull [] @NotNull [] matrix) {
        if (matrix.length == 0) {
            throw new IllegalArgumentException();
        }
        for (var i = 0; i < matrix.length; ++i) {
            if (matrix[i].length != matrix.length) {
                throw new IllegalArgumentException();
            }
            if (matrix[i][i] != -1) {
                throw new IllegalArgumentException();
            }
            for (var j = 0; j < i; ++j) {
                if (matrix[j][i] != matrix[i][j] || (matrix[j][i] != -1 && matrix[i][j] < EPSILON)) {
                    throw new IllegalArgumentException();
                }
            }
        }
        return -1;
    }
}
